### ISSUE TITLE
Incorrect Result in Nested Joins with Citus Tables

### ISSUE THREAD

**Issue by halilozanakgul** [citusdata/citus#6636]
After creating tables like this:
```SQL
DROP TABLE IF EXISTS tbl_dist1;
CREATE TABLE tbl_dist1(id int);
SELECT create_distributed_table('tbl_dist1','id');

DROP TABLE IF EXISTS tbl_ref1;
CREATE TABLE tbl_ref1(id int);
SELECT create_reference_table('tbl_ref1');

INSERT INTO tbl_dist1 SELECT i FROM generate_series(0,10) i;
INSERT INTO tbl_ref1 SELECT i FROM generate_series(0,10) i;
```
if you run
```SQL
WITH cte_0 AS (
    SELECT id FROM tbl_dist1 WHERE id IN (
        SELECT id FROM tbl_ref1
        LEFT JOIN tbl_dist1 USING (id)
    )
)
SELECT count(id) FROM tbl_dist1
RIGHT JOIN (
    SELECT table_5.id FROM (
        SELECT id FROM cte_0 LIMIT 0
    ) AS table_5
    RIGHT JOIN tbl_dist1 USING (id)
) AS table_4 USING (id);
```

the output is:
```SQL
 count
-------
    11
(1 row)
```

but if you remove `create_distributed_table` and `create_reference_table` queries while creating the tables you get:
```SQL
 count
-------
     0
(1 row)
```

**Comment by halilozanakgul** [comment]
Got another example that looks very similar
```SQL
DROP TABLE IF EXISTS dist0;
CREATE TABLE dist0(id int);
SELECT create_distributed_table('dist0','id');

DROP TABLE IF EXISTS dist1;
CREATE TABLE dist1(id int);
SELECT create_distributed_table('dist1','id');

INSERT INTO dist0 SELECT i FROM generate_series(1005,1025) i;
INSERT INTO dist1 SELECT i FROM generate_series(1015,1035) i;
```

```SQL
WITH cte_0 AS (
    SELECT id FROM dist0
    RIGHT JOIN dist0 AS table_1 USING (id)
    ORDER BY id
)
SELECT avg(avgsub.id) FROM (
    SELECT table_2.id FROM (
        SELECT table_3.id FROM (
            SELECT table_5.id FROM cte_0 AS table_5, dist1
        ) AS table_3 INNER JOIN dist1 USING (id)
    ) AS table_2 FULL JOIN dist0 USING (id)
) AS avgsub;
```

**Comment by halilozanakgul** [comment]
One more
```SQL
DROP TABLE IF EXISTS dist0;
CREATE TABLE dist0(id int);
SELECT create_distributed_table('dist0','id');

DROP TABLE IF EXISTS dist1;
CREATE TABLE dist1(id int);
SELECT create_distributed_table('dist1','id');

INSERT INTO dist0 SELECT i FROM generate_series(0,10) i;
INSERT INTO dist0 SELECT * FROM dist0 ORDER BY id LIMIT 1;

INSERT INTO dist1 SELECT i FROM generate_series(0,10) i;
INSERT INTO dist1 SELECT * FROM dist1 ORDER BY id LIMIT 1;
```

```SQL
SELECT avg(avgsub.id) FROM (
    SELECT table_0.id FROM (
        SELECT table_1.id FROM (
            SELECT table_2.id FROM (
                SELECT table_3.id FROM (
                    SELECT table_4.id FROM dist0 AS table_4
                    LEFT JOIN dist1 AS table_5 USING (id)
                ) AS table_3 INNER JOIN dist0 AS table_6 USING (id)
            ) AS table_2 WHERE table_2.id < 10 ORDER BY id LIMIT 47
        ) AS table_1 RIGHT JOIN dist0 AS table_7 USING (id)
    ) AS table_0 RIGHT JOIN dist1 AS table_8 USING (id)
) AS avgsub;

**Comment by onurctirtir** [comment]
I feel like there is something going wrong around subplan id assignment / lifetime of intermediate results:

```sql
WITH cte_0 AS (
    SELECT table_0.id FROM tbl_dist1 AS table_0 LEFT JOIN tbl_dist1 AS table_1 USING (id) ORDER BY id LIMIT 5
)
SELECT avg(avgsub.id) FROM (
    SELECT table_4.id FROM (
        SELECT table_5.id FROM (
            SELECT table_6.id FROM cte_0 AS table_6
        ) AS table_5
        INNER JOIN tbl_dist1 USING (id) INNER JOIN tbl_dist1 AS table_9 USING (id)
    ) AS table_4 FULL JOIN tbl_dist1 USING (id)
) AS avgsub;
WARNING:  Query could not find the intermediate result file "119_2", it was mostly likely deleted due to an error in a parallel process within the same distributed transaction
```

**Comment by onurctirtir** [comment]
> I feel like there is something going wrong around subplan id assignment / lifetime of intermediate results:
> 
> ```sql
> SELECT t1.id
> FROM (
>     SELECT t2.id
>     FROM (
>         SELECT t0.id
>         FROM tbl_dist1 t0
>         LIMIT 5        
>     ) AS t2
>     INNER JOIN tbl_dist1 AS t3 USING (id)
> ) AS t1
> FULL JOIN tbl_dist1 t4 USING (id);
> WARNING:  Query could not find the intermediate result file "143_2", it was mostly likely deleted due to an error in a parallel process within the same distributed transaction
> ```

Due to LIMIT clause used in the innermost query, the query becomes the following after replacing that part of the query with intermediate result:
```sql
SELECT t1.id
FROM (
    (
        SELECT t2.id
        FROM (
            (
                SELECT intermediate_result.id
                FROM read_intermediate_result('143_1'::text, 'binary'::citus_copy_format) intermediate_result(id integer)
            ) t2
            JOIN public.tbl_dist1 t3 USING (id)
        )
    )
    t1 FULL JOIN public.tbl_dist1 t4 USING (id)
)
```

And later on, we decide recursively planning `t4` --by using a subplan with id 143_2-- because we decide that the other side of the FULL JOIN is recurring (just because it's a subquery ?, anyway). And the warning states that the subplan with id 143_2 is missing.

**Comment by aykut-bozkurt** [comment]
We found 3 different cases which causes recursive planner passes the query multiple times in PR #6650. We throw error if we detect multipass of recursive planner. Proper fix needs 3 cases to be handled properly.
1. Sublink in WHERE clause is planned at second pass after we recursively planned a distributed table at the first pass. Fixed by PR #6657.
2. Local-distributed joins are recursively planned at both the first and the second pass. Issue #6659.
3. Some parts of the query is considered to be noncolocated at the second pass as we do not generate attribute equivalances between nondistributed and distributed tables. Issue #6653
